core/blockchain.go

```diff
@@ -616,6 +616,132 @@ func (bc *BlockChain) SetSafe(header *types.Header) {
 	}
 }
 
+// rewindPathHead implements the logic of rewindHead in the context of hash scheme.
+func (bc *BlockChain) rewindHashHead(root common.Hash) (*types.Header, uint64) {
+	var (
+		limit      uint64                             // The oldest block that will be searched for this rewinding
+		beyondRoot = root == common.Hash{}            // Flag whether we're beyond the requested root (no root, always true)
+		pivot      = rawdb.ReadLastPivotNumber(bc.db) // Associated block number of pivot point state
+		headBlock  = bc.CurrentBlock()                // Head block of the chain
+		rootNumber uint64                             // Associated block number of requested root
+	)
+	// The oldest block to be searched is determined by the pivot block or a soft
+	// finality threshold. The rationale behind this is as follows:
+	//
+	// - Snap sync is selected if the pivot block is available. The earliest available
+	//   state is the pivot block itself, so there is no sense in going further back.
+	//
+	// - Full sync is selected if the pivot block does not exist. The hash database
+	//   periodically flushes the state to disk, and the soft finality threshold is
+	//   considered sufficient to find a persistent state, even for the testnet. It
+	//   might be not enough for a chain that is nearly empty. In the worst case,
+	//   the entire chain is reset to genesis, and snap sync is re-enabled on top,
+	//   which is still acceptable.
+	if pivot != nil {
+		limit = *pivot
+	} else if headBlock.Number.Uint64() > params.FullImmutabilityThreshold {
+		limit = headBlock.Number.Uint64() - params.FullImmutabilityThreshold
+	}
+	for {
+		// If a root threshold was requested but not yet crossed, check
+		if !beyondRoot && headBlock.Root == root {
+			beyondRoot, rootNumber = true, headBlock.Number.Uint64()
+		}
+		// If the associated state is not reachable, continue searching
+		// backwards until an available state is found.
+		if !bc.HasState(headBlock.Root) {
+			log.Trace("Block state missing, rewinding further", "number", headBlock.Number, "hash", headBlock.Hash())
+
+			// If search limit is reached, return the genesis block as
+			// the new chain head.
+			if headBlock.Number.Uint64() < limit {
+				log.Info("Rewinding limit is reached, aiming genesis", "number", headBlock.Number, "hash", headBlock.Hash(), "limit", limit)
```

Comment:
"Rewinding limit is reached, aiming genesis" -> "Rewinding limit reached, resetting to genesis"